cert-manager installation using argocd (#1350)
cert-manager/website: content/docs/installation/continuous-deployment-and-gitops.md

```bash
kubectl logs -n cert-manager -l app.kubernetes.io/instance=cert-manager --prefix --all-containers
```

## Using ArgoCD
Argo CD is a declarative, GitOps continuous delivery tool for Kubernetes.

### Pre-requisites
Ensure the following are in place before proceeding:
- A Kubernetes cluster
- ArgoCD deployed on the Kubernetes cluster: [installation guide](https://argo-cd.readthedocs.io/en/stable/getting_started/)
- Optional: A GitOps repository connected with ArgoCD: [setup guide](https://argo-cd.readthedocs.io/en/stable/user-guide/private-repositories/)

### Setting up cert-manager

> 1. Create an ArgoCD Application manifest file with the provided configuration to set up cert-manager.

Reviewer: ```suggestion
1. Create an [ArgoCD Application manifest file](link-to-reference-documentation) with the provided configuration to set up cert-manager.
```

It will be useful to link to the documentation so that the reader can learn about the fields of this custom resource.

Reviewer 2: @wallrj this is not my code/fix BUT i'm currently affected by this.
I will pull this branch and make the suggested edits after my comments below are accepted or better suggestions are made.

Reviewer: Thanks. I don't have permission to push to @shahkv95 's branch, so I've created my own.
 * https://github.com/cert-manager/website/pull/1447

    ```yaml
    apiVersion: argoproj.io/v1alpha1
    kind: Application
    metadata:
      name: cert-manager
      namespace: argocd
      annotations:
        argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
      finalizers:

>         - resources-finalizer.argocd.argoproj.io

Reviewer 2: These are part of the (Argo) Application resource. Linking to the official documentation for this resource in the Argo documentation should be enough.
I think we want people to go read the official docs vs just copying and pasting from this docs.

    spec:
      destination:
        namespace: cert-manager

>         server: https://kubernetes.default.svc

Reviewer: I'm surprised that you have to declare the URL of the Kubernetes API server. Is it strictly necessary? If not, I suggest removing it to avoid distracting the reader.

      project: default
      source:
        chart: cert-manager
        repoURL: https://charts.jetstack.io

>         targetRevision: 1.10.1

Reviewer: Use the latest cert-manager version: 1.13.2 at time of writing.

Reviewer: Replaced with variable. See:
 * https://github.com/cert-manager/website/pull/1433

        helm:
          values: |
            installCRDs: true
      syncPolicy:
        automated:
          prune: true

>           selfHeal: true

Reviewer: Add YAML comments above these fields explaining why these options are needed.

Reviewer 2: These are part of the (Argo) Application resource. Linking to the official documentation for this resource in the Argo documentation should be enough.
I think we want people to go read the official docs vs just copying and pasting from this docs.

        syncOptions:
          - CreateNamespace=true
    ```
2. Commit the manifest file and sync the changes in ArgoCD. If a GitOps repository is not set up, use `kubectl apply -f <above-file-path>` to apply the manifest [installation guide for kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl).

> 3. ArgoCD will synchronize the Desired manifest and deploy cert-manager on Kubernetes based on the configuration provided.

Reviewer 3: ```suggestion
3. ArgoCD will synchronize the `DESIRED MANIFEST` and deploy cert-manager on Kubernetes based on the provided configuration.
```

### Troubleshooting

#### Scenario 1:

> Out-of-sync cert-manager in AKS(Azure Kubernetes Service) cluster

Reviewer 3: ```suggestion
Out-of-sync cert-manager in AKS (Azure Kubernetes Service) cluster
```

> Out-of-sync cert-manager in the AKS (Azure Kubernetes Service) cluster

Reviewer 3: That's what I'd expect

##### Issue: 

> Cert-manager in the AKS cluster remains out-of-sync due to discrepancies between the Desired and Live manifest files.

Reviewer 3: ```suggestion
Cert-manager in the AKS cluster remains out-of-sync due to discrepancies between the `DESIRED MANIFEST` and `LIVE MANIFEST` files.
```

##### Potential Reasons
Multiple factors could cause the OutOfSync issue; refer to [ArgoCD documentation](https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#diffing-customization) for potential causes.

##### Example configuration differences

> Below configurations are observed to be present in the Live manifest but not in the Desired manifest file.

Reviewer 3: ```suggestion
Below configurations are observed to be present in the `LIVE MANIFEST` but not in the `DESIRED MANIFEST`.
```

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
...
webhooks:
- admissionReviewVersions:
  namespaceSelector:
    matchExpressions:
        ...
        ...
        - key: control-plane
          operator: NotIn
          values:
            - 'true'
        - key: kubernetes.azure.com/managedby
          operator: NotIn
          values:
            - aks
```

##### Root Cause Analysis

> The discrepancy is rooted in how AKS manages admission controllers to protect internal services in the kube-system namespace. More details can be found [here](https://learn.microsoft.com/en-us/azure/aks/faq#can-admission-controller-webhooks-impact-kube-system-and-internal-aks-namespaces)

Reviewer 3: ```suggestion
The discrepancy stems from how AKS manages admission controllers to protect internal services in the `kube-system` namespace. More details can be found in [Frequently asked questions about Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/faq#can-admission-controller-webhooks-impact-kube-system-and-internal-aks-namespaces)
```

##### Suggested Fix

> It is also possible to ignore differences from fields owned by specific managers defined in `metadata.managedFields` in live resources. More details can be found [here](https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#application-level-configuration)

Reviewer 3: ```suggestion
It is also possible to ignore differences from fields owned by specific managers defined in `metadata.managedFields` in live resources. More details can be found in [(ArgoCD) Diffing Customization](https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#application-level-configuration)
```

      - .webhooks[].namespaceSelector.matchExpressions[] | select(.key == "kubernetes.azure.com/managedby")
```

In that case, the updated cert-manager manifest would be as follows:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: cert-manager
  namespace: argocd
  annotations:
    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  destination:
    namespace: cert-manager
    server: https://kubernetes.default.svc
  project: default
  source:
    chart: cert-manager
    repoURL: https://charts.jetstack.io
    targetRevision: 1.10.1
    helm:
      values: |
        installCRDs: true
  ignoreDifferences:
  - group: admissionregistration.k8s.io
    kind: ValidatingWebhookConfiguration
    name: cert-manager-webhook
    jqPathExpressions:
      - .webhooks[].namespaceSelector.matchExpressions[] | select(.key == "control-plane")
      - .webhooks[].namespaceSelector.matchExpressions[] | select(.key == "kubernetes.azure.com/managedby")
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```

> Once ArgoCD syncs the updated manifest, the differences due to above 2 keys would be ignored and cert-manager would be in complete synchronization state.

Reviewer 3: ```suggestion
Once ArgoCD syncs the updated manifest, the differences due to above two keys will be ignored and cert-manager would be in complete synchronization state.
```
